### PR TITLE
Deployment related tasks should use explodedAppDirectory. Discover SystemServiceServlet either by servlet-name or servlet-class.

### PR DESCRIPTION
**Pull request by arskov**
This pull request contains 2 major improvements. The plugin will allow to discover SystemServiceServlet either by servlet-name or servlet-class. Useful if you extend SystemServiceServlet with custom logic, for example logging, and deployment related tasks (appengineUpdateCron, appengineUpdateQueues, etc) should use explodedAppDirectory instead of source webAppSourceDirectory. It is very useful, for example, if you use placeholders in appengine-web.xml that are being filled from command line parameters. Relates to #193.

### CODE REVIEW

File: builder-model/.gitignore [hunk 1]
```diff
@@ -0,0 +1,11 @@
+build
```
Comment (loosebazooka) [inline builder-model/.gitignore]: I don't think this project should load as it's own project.  It's part of the greater gradle build.  Anyway, the parent .gitignore should cover all this.

File: src/main/groovy/com/google/appengine/AppEnginePlugin.groovy [hunk 2]
```diff
@@ -121,33 +121,44 @@ class AppEnginePlugin implements Plugin<Project> {
...
+    private File endpointsClientLibDirectory
+    private File endpointsExpandedSrcDirectory
+    
+    private AppEnginePluginExtension appEnginePluginExtension
+
     @Override
     void apply(Project project) {
         checkGradleVersion(project)
-        registry.register(new AppEngineToolingBuilderModel())
+        this.registry.register(new AppEngineToolingBuilderModel())
```
Comment (loosebazooka) [inline src/main/groovy/com/google/appengine/AppEnginePlugin.groovy]: I don't think "this" is necessary here. There's no clash with local variables.

File: src/main/groovy/com/google/appengine/AppEnginePlugin.groovy [hunk 3]
```diff
@@ -245,28 +256,32 @@ class AppEnginePlugin implements Plugin<Project> {
         }
     }
 
-    private void configureAppConfig(Project project, AppEnginePluginExtension appEnginePluginExtension) {
+    private void configureWorkingDirectory(Project project) {
+        AppConfigTaskTemplate.metaClass.getWorkingDirectory = { -> this.explodedAppDirectory }
```
Comment (loosebazooka) [inline src/main/groovy/com/google/appengine/AppEnginePlugin.groovy]: Use of "this" in the whole file is probably unnecessary.

Comment (loosebazooka) [inline src/main/groovy/com/google/appengine/AppEnginePlugin.groovy]: I would prefer if "workingDirectory" was defined in WebAppDirTask instead of via a metaClass here, we don't really need to complicate the code if we know what we want to do.

File: src/main/groovy/com/google/appengine/task/endpoints/WebXmlProcessing.groovy [hunk 4]
```diff
@@ -36,7 +36,28 @@ class WebXmlProcessing {
     }
 
     private static NodeChild getSystemServiceServlet(GPathResult root) {
-        def servlets = root."servlet".findAll { it."servlet-class".text().trim() == "com.google.api.server.spi.SystemServiceServlet" ? it : null }
+        def servlets = root."servlet".findAll { servlet ->
```
Comment (loosebazooka) [inline src/main/groovy/com/google/appengine/task/endpoints/WebXmlProcessing.groovy]: Can you give me an example of when this is necessary? I'm not sure I follow.

**Comment by loosebazooka** [comment]
I would prefer not to override the default behavior abruptly, I'm not sure existing users wont experience breakages or unexpected behavior.  Perhaps a system property that says "use.output.webapp.dir" that sets webAppDirectory in WebAppDirTask to the exploded app webapp dir instead of the source folder? And then we start to slowly change the default behavior.

File: src/main/groovy/com/google/appengine/task/endpoints/WebXmlProcessing.groovy (hunk 4 above)
Comment (loosebazooka) [inline src/main/groovy/com/google/appengine/task/endpoints/WebXmlProcessing.groovy]: Also could this go in it's own PR as it is unrelated to the current theme

Comment (ArseniKavalchuk) [inline src/main/groovy/com/google/appengine/task/endpoints/WebXmlProcessing.groovy]: For example I'm using custom overridden implementation of SystemServiceServlet
```xml
	<servlet>
		<servlet-name>SystemServiceServlet</servlet-name>
		<servlet-class>com.severn.common.http.WrappedSystemServiceServlet</servlet-class>
		<init-param>
			<param-name>services</param-name>
			<param-value></param-value>
		</init-param>
        <init-param>
            <param-name>logRequest</param-name>
            <param-value>false</param-value>
        </init-param>
        <init-param>
            <param-name>logResponse</param-name>
            <param-value>true</param-value>
        </init-param>
        <init-param>
            <param-name>endpointsPath</param-name>
            <param-value>/_ah/spi/com.severn</param-value>
        </init-param>
	</servlet>
    <servlet-mapping>
        <servlet-name>SystemServiceServlet</servlet-name>
        <url-pattern>/_ah/spi/*</url-pattern>
    </servlet-mapping>
```
If we stick only to servlet class, it won't allow to find endpoint classes and fails to generate discovery docs in cases like above. Servlet name is more flexible, however, is not perfect.
Related TODOs are: 
1. Find servlet name by its mapping `/_ah/spi/*`, and then find servlet declaration, then get `services` init param.
2. Implement a search of endpoint services by @Api annotations like maven plugin does and put them to the services param  

Agree, it should go in another PR.

File: src/main/groovy/com/google/appengine/AppEnginePlugin.groovy (hunk 3 above)
Comment (ArseniKavalchuk) [inline src/main/groovy/com/google/appengine/AppEnginePlugin.groovy]: In regard to `this`, for example, the property `explodedSdkDirectory` clashes with getter method `getExplodedSdkDirectory` and results in `StackOverflowError` in `appengineDownloadSdk` task. That was the reason to guard all properties instead of digging to the each particular case.

Comment (ArseniKavalchuk) [inline src/main/groovy/com/google/appengine/AppEnginePlugin.groovy]: In regard to `workingDirectory`, I agree, looks ugly. What do you think about DSL config flag that will point which directory to use, like:
```
appengine {
    useOutputWebappDir=true
}
```
instead of system property "use.output.webapp.dir"?
Probably, it doesn't reflect the actual meaning, requires documentation, or may be different name.

**Comment by ArseniKavalchuk** [comment]
Thanks, for the review. See inline comments about particular cases. Let's agree about how we supply the flag "use.output.webapp.dir" about which directory to use and I'll implement.

File: src/main/groovy/com/google/appengine/AppEnginePlugin.groovy (hunk 3 above)
Comment (loosebazooka) [inline src/main/groovy/com/google/appengine/AppEnginePlugin.groovy]: So I didn't think there would be automatic getter/setter creation for private members in groovy, but if there's some naming ambiguity maybe we should change things around instead of letting the ambiguity persist.  Perhaps changing the static methods from "get" to something that doesn't clash here.

Comment (loosebazooka) [inline src/main/groovy/com/google/appengine/AppEnginePlugin.groovy]: The system property vs DSL might be a more difficult question. I think it's a little bit of a preference thing, I suggested the system property so it can be kind of a temporary holdover while we transition everyone to using the output dir by default, without having to maintain both code paths after the transition is complete.

File: src/main/groovy/com/google/appengine/task/endpoints/WebXmlProcessing.groovy (hunk 4 above)
Comment (loosebazooka) [inline src/main/groovy/com/google/appengine/task/endpoints/WebXmlProcessing.groovy]: Ah makes sense.  So there is another option here where you can specify `serviceClasses` instead of relying on auto detection.

Lets move this to another PR though.

**Comment by ArseniKavalchuk** [comment]
I signed Contributor License Agreement
